File: Lab3/src/c-matrix.py
# Implement a two-class confusion matrix using your own code with computing  FP, FN, TP, TN, AC (accuracy), P (precision), R (recall).

import pandas as pd
from sklearn.datasets import load_wine
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn import metrics

raw_data = load_wine()

df = pd.DataFrame(data=raw_data['data'], columns=raw_data['feature_names'])

df['target'] = raw_data['target']
df['class'] = df['target'].map(lambda i: raw_data['target_names'][i])

# convert to binary classifier: in class_0 (class_0), not in class_0 (class_1)
df.loc[df['target'] == 2, 'target'] = 1
df.loc[df['class'] == 'class_2', 'class'] = 'class_1'

df.to_csv('dat/wine-data.csv')
predictors = df.columns[:len(df.columns) - 2]
X = df[predictors]
y = df['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

clf = DecisionTreeClassifier()

# train Decision Tree Classifer
clf = clf.fit(X_train, y_train)

# predict the response for test dataset
y_pred = clf.predict(X_test)

a, b, c, d = metrics.confusion_matrix(y_test, y_pred).ravel()

# accuracy
AC = (a + d)/(a + b + c + d)
print("AC:", AC)

# true positive rate (recall)
TP = d/(c + d)
print("TP (R):", TP)

# false positive rate
FP = b/(a + b)
print("FP:", FP)

# true negative rate
TN = a/(a + b)
print("TN:", TN)

# false negative rate
FN = c/(c + d)
print("FN:", FN)

# precision
P = d/(b + d)
print("P:", P)

File: Lab3/src/c-matrix.R
# Implement a two-class confusion matrix using your own code with 
# computing  FP, FN, TP, TN, AC (accuracy), P (precision), R (recall).

# decision tree with rpart
library(rpart)
library(rpart.plot)
library(caret)

# Set working directory
setwd("~/git/data-mining/Lab3/src")

# load dataframe
wine_df = read.csv("../dat/wine-data.csv", header=TRUE)

# remove unneccesay columns
wine_df = subset(wine_df, select = -c(X, class))

# split into training/testing sets
set.seed(509)

# train percentage
train <- sample(1:nrow(wine_df), size = ceiling(0.8*nrow(wine_df)), replace = FALSE)

training_df <- wine_df[train,]
testing_df <- wine_df[-train,]

# grow tree
my_tree <- rpart(target~., data = training_df, method = "class")

# visualize tree
rpart.plot(my_tree, nn = TRUE)


# testing model
predictions <-  predict(object = my_tree, testing_df[1:length(testing_df) - 1], type = "class")

# create a confusion matrix
my_table <- table(testing_df$target, predictions)

a <- my_table[1, "0"]
b <- my_table[1, "1"]
c <- my_table[2, "0"]
d <- my_table[2, "1"]

# final calculations based on confusion matrix values

# accuracy
AC = (a + d)/(a + b + c + d)
print(paste0("AC: ", formatC(100*AC, digits = 2, format = "f"), "%"))

# true positive rate (recall)
TP = d/(c + d)
print(paste0("TP (R): ", TP))

# false positive rate
FP = b/(a + b)
print(paste0("FP: ", FP))

# true negative rate
TN = a/(a + b)
print(paste0("TN: ", TN))

# false negative rate
FN = c/(c + d)
print(paste0("FN: ", FN))

# precision
P = d/(b + d)
print(paste0("P: ", P))



